include/envoy/server/transport_socket_config.h

```diff
@@ -0,0 +1,66 @@
+#pragma once
+
+#include <string>
+
+#include "envoy/network/transport_socket.h"
+#include "envoy/ssl/context_manager.h"
+
+#include "common/protobuf/protobuf.h"
+
+namespace Envoy {
+namespace Server {
+namespace Configuration {
+
+/**
+ * Context passed to transport socket factory to access server resources.
+ */
+class TransportSocketFactoryContext {
+public:
+  virtual ~TransportSocketFactoryContext() {}
+
+  /**
+   * @return Ssl::ContextManager& the SSL context manager
+   */
+  virtual Ssl::ContextManager& sslContextManager() PURE;
+
+  /**
+   * @return Stats::Scope& the tansprot socket's stats scope.
+   */
+  virtual Stats::Scope& statsScope() const PURE;
+};
+
+class TransportSocketConfigFactory {
+public:
+  virtual ~TransportSocketConfigFactory() {}
+
+  /**
+   * Create a particular transport socket factory implementation. If the implementation is unable
+   * to produce a factory with the provided parameters, it should throw an EnvoyException. The
+   * returned TransportSocketFactoryPtr should not be nullptr.
+   * @param config supplies the config for the transport socket implementation.
+   * @param context supplies the transport socket's context.
+   * @return Network::TransportSocketFactoryPtr the transport socket factory instance.
+   */
+  virtual Network::TransportSocketFactoryPtr
+  createTransportSocketFactory(const Protobuf::Message& config,
+                               TransportSocketFactoryContext& context) PURE;
+
+  /**
+   * @return ProtobufTypes::MessagePtr create empty config proto message. The transport socket
+   *         config, which arrives in an opaque google.protobuf.Struct message, will be converted
+   *         to JSON and then parsed into this empty proto.
+   */
+  virtual ProtobufTypes::MessagePtr createEmptyConfigProto() PURE;
```

Comment:
Where is this used?

Reply:
in `translateToFactoryConfig`:
https://github.com/envoyproxy/envoy/blob/master/source/common/config/utility.h#L208